logstash-core/lib/logstash/util/substitution_variables.rb

```diff
@@ -31,26 +31,11 @@ def deep_replace(value)
   def replace_placeholders(value)
     return value unless value.is_a?(String)
 
-    value.gsub(SUBSTITUTION_PLACEHOLDER_REGEX) do |placeholder|
-      # Note: Ruby docs claim[1] Regexp.last_match is thread-local and scoped to
-      # the call, so this should be thread-safe.
-      #
-      # [1] http://ruby-doc.org/core-2.1.1/Regexp.html#method-c-last_match
-      name = Regexp.last_match(:name)
-      default = Regexp.last_match(:default)
-      logger.debug("Replacing `#{placeholder}` with actual value")
-
-      #check the secret store if it exists
-      secret_store = LogStash::Util::SecretStore.get_if_exists
-      replacement = secret_store.nil? ? nil : secret_store.retrieveSecret(LogStash::Util::SecretStore.get_store_id(name))
-      #check the environment
-      replacement = ENV.fetch(name, default) if replacement.nil?
-      if replacement.nil?
-        raise LogStash::ConfigurationError, "Cannot evaluate `#{placeholder}`. Replacement variable `#{name}` is not defined in a Logstash secret store " +
-            "or as an Environment entry and there is no default value given."
-      end
-      replacement.to_s
-    end
+    org.logstash.common.SubstitutionVariables.replacePlaceholders(
+        value, ENV, LogStash::Util::SecretStore.get_if_exists
+    );
+  rescue org.logstash.common.SubstitutionVariables::MissingSubstitutionVariableError => e
```

Comment:
@andrewvc are you sure this is caught here? Instinctively I'd say now actually. Maybe simply throw that `ConfigurationError` on the Java side in the first place?

Syntax for that would be:

```java
                throw RubyUtil.RUBY.newRaiseException(
                    RubyUtil.CONFIGURATION_ERROR_CLASS,
                    "Your Message"
                );
```

:)

Reply:
What do you mean by "am I sure that is caught?" As in, is the thrown exception caught, or is the `rescue` block here doing its work?

I'd prefer not to wrap this with a JRuby ext wrapper that throws ruby exceptions unless we need it. 

This does work correctly today and is caught by https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/util/settings_helper.rb#L41

I verified this manually.

Reply:
@andrewvc yea that's what I meant, all good then :)